app/quadraticlands/views.py

```diff
@@ -295,3 +302,145 @@ def error(request, code):
+
+
+@login_required
+def mission_diplomacy_room(request, uuid, name):
+    # get the game
+    try:
+        game = Game.objects.get(uuid=uuid)
+    except:
+        raise Http404
+    return mission_diplomacy_room_helper(request, game)
+
+def mission_diplomacy_room_helper(request, game):
+
+    # handle invites
+    users_players = GamePlayer.objects.filter(active=True, profile=request.user.profile)
+    users_games = Game.objects.filter(pk__in=users_players.values_list("game"))
+    users_player =  users_players.first()
+    if game not in users_games:
+        if request.user.is_authenticated:
+
+            # too many user condition
+            if game.active_players.count() >= max_players_per_game:
+                messages.info(
+                    request,
+                    f'There are already {max_players_per_game} in the game.  Wait for someone to leave + try again.'
+                )
+                return mission_diplomacy_helper(request)
+
+            # show invite
+            if request.GET.get('join'):
+                game.add_player(request.user.profile.handle)
+            else:
+                return mission_diplomacy_helper(request, invited_to_game=game)
+        else:
+            # logged out condition
+            messages.info(
+                request,
+                f'Please login to join this game.'
+            )
+
+    # in game experience
+    is_member = game.is_active_player(request.user.profile.handle)
+    is_admin = game.players.filter(admin=True, profile=request.user.profile).exists()
+    if is_admin:
+        if request.GET.get('remove'):
+            remove_this_fool = request.GET.get('remove')
+            for player in game.players.filter(active=True, profile__handle=remove_this_fool):
+                game.remove_player(player.profile.handle)
+                messages.info(
+                    request,
+                    f'{remove_this_fool} has been removed'
+                )
+    
+    # delete game
+    if is_member and request.GET.get('delete'):
+        game.remove_player(request.user.profile.handle)
+        messages.info(
+            request,
+            f'You have left this game.'
+        )
+        return redirect('/quadraticlands/mission/diplomacy')
+
+    # chat in game
+    if is_member and request.POST.get('chat'):
+        game.chat(request.user.profile.handle, request.POST.get('chat'))
+
+    # make a move
+    if is_member and request.POST.get('signature'):
+        moves = json.loads(request.POST.get('package'))
+        signature = request.POST.get('signature')
+        recipient_address = Web3.toChecksumAddress(moves['account'])
+        web3 = get_web3('mainnet')
+        gtc = web3.eth.contract(address=Web3.toChecksumAddress('0xde30da39c46104798bb5aa3fe8b9e0e1f348163f'), abi=erc20_abi)
+        balance = gtc.functions.balanceOf(recipient_address).call()
+        import ipdb; ipdb.set_trace()
```

Comment:
hey @thelostone-mc heres where i need help Validating the signed message upon vouch. I can’t find the right function for this since we use an old web3py


we want to do two things on backend.

1. validate the balance is what the frontend says 
2. validate the message was signed by the address the frontend claims

Reply:
This is what the frontend sends to the backend

![image](https://user-images.githubusercontent.com/5358146/123641230-aa803400-d83f-11eb-8ade-3283fdac75a4.png)

This is the code which would help in both balance check and validate signer

```python
from web3 import Web3
from dashboard.utils import get_web3
from dashboard.abi import erc20_abi
from eth_account.messages import defunct_hash_message

web3 = get_web3('mainnet')

# Balance
gtc = web3.eth.contract(address=Web3.toChecksumAddress('0xde30da39c46104798bb5aa3fe8b9e0e1f348163f'), abi=erc20_abi)
recipient_address = '0x1bCD46B724fD4C08995CEC46ffd51bD45feDE200'
balance = gtc.functions.balanceOf(recipient_address).call()

print(balance)

# Validate  Sign
sign = '0x7f2c5571b08fc56697ee07de688a3c1c9fd980cadba3f85cc4f7cacb7e0bb22422a7056ab31ec303b2696b04a41f0b0e1ba90f909af06637f2cb73fe9027c75f1b'

msg = '{"votes":[{"userid":"601","username":"thelostone-mc","value":0},{"userid":"64423","username":"shitzu-mc","value":"1"}],"balance":"0","account":"0x1bCD46B724fD4C08995CEC46ffd51bD45feDE200"}'
message_hash = defunct_hash_message(text=msg)

signerAddress = web3.eth.account.recoverHash(message_hash, signature=sign)

print(recipient_address == signerAddress)
```

Here is screenshot of it on shell working 

<img width="1421" alt="Screenshot 2021-06-28 at 6 48 35 PM" src="https://user-images.githubusercontent.com/5358146/123643143-b836b900-d841-11eb-938d-10633fbbd75b.png">

Reply:
376d8ebe0bc9fd123ca436179832682110b13c00